cr-examples/samples/src/main/java/oracle/code/samples/MathOptimizer.java

```diff
@@ -0,0 +1,281 @@
+
+                    // Analyse second operand of the Math.pow(x, y).
+                    // if the x == 2, and both are integers, then we can optimize the function using bitwise operations
+                    // pow(2, y) replace with (1 << y)
+                    Value operand = operands.getFirst();  // obtain the first parameter
+                    // inspect if the base (as in pow(base, exp) is value 2
+                    boolean canApplyBitShift = inspectParameterRecursive(operand, 2);
+                    if (canApplyBitShift) {
+                        // We also need to inspect types. We can apply this optimization
+                        // if the exp type is also an integer.
+                        boolean isIntType = analyseType(operands.get(1), JavaType.INT);
+                        if (!isIntType) {
+                            canApplyBitShift = false;
+                        }
+                    }
+
+                    // If the conditions to apply the first optimization failed, we try the second optimization
+                    // if types are not int, and base is not 2.
+                    // pow(x, 2) => replace with x * x
+                    boolean canApplyMultiplication = false;
+                    if (!canApplyBitShift) {
+                        // inspect if exp (as in pow(base, exp) is value 2
+                        canApplyMultiplication = inspectParameterRecursive(operands.get(1), 2);
+                    }
+
+                    if (canApplyBitShift) {
+                        // Narrow type from DOUBLE to INT for the input parameter of the new function.
+                        Op.Result op2 = blockBuilder.op(JavaOp.conv(JavaType.INT, operands.get(1)));
+                        List<Value> newOperandList = new ArrayList<>();
+                        newOperandList.add(op2);
+
+                        // Create a new invoke with the optimised method
+                        JavaOp.InvokeOp newInvoke = JavaOp.invoke(MY_SHIFT_FUNCTION, newOperandList);
+                        // Copy the original location info to the new invoke
+                        newInvoke.setLocation(invokeOp.location());
+
+                        // Replace the invoke node with the new optimized invoke
+                        Op.Result newResult = blockBuilder.op(newInvoke);
+                        blockBuilder.context().mapValue(invokeOp.result(), newResult);
+
+                    } else if (canApplyMultiplication) {
+                        // Adapt the parameters to the new function. We only need the first
+                        // parameter from the initial parameter list  - pow(x, 2) -
+                        // Thus, we copy parameter x into a new list and pass it to the new
+                        // invoke function.
+                        List<Value> newOperandList2 = new ArrayList<>();
+                        newOperandList2.add(operands.get(0));
```

Comment:
You don't need to create list (or a sub list). The `Java.invoke` method is overloaded to accept a var args or a list. In this case we can use the former and just pass `operands.get(0)` directly, or more clearly `operands.getFirst()`.